prob() with rand() is the same as a plain prob()

Szunti opened the pull request:
Isn't it?

RemieRichards commented:
What?
I assume that would be evaluated as
```
X = rand(1,100)
Prob(X)
```

xxalpha commented:
It's not the same, it's choosing a probability at random. Might be too much random.

Szunti commented:
Yes, it's

X = random number from 1 to 100 with uniform distribution

so the probability of every value is 1/100

Then for every X the conditional probability of Prob(X) is X/100

So the probability of Prob(rand(1,100)) is 
1/100 * (1/100 + 2/100 + ... 100/100) = 100*101/2 /100/100 = 55/100 which is Prob(55)

RemieRichards commented:
it doesn't do the prob() operation on each possible value.
it finds a value between 1 and 100, say, 22, and then prob()s that.

Szunti commented:
I never said it does. What I'm trying to say is that probability of a well known random event is just a probability.

Menshin commented:
100*101 / 2 = 5050 and 5050 / 100 = 50.5
Hence, you would have to test prob(50.5).

Also everything above is considering we're working with integers (so, _discrete_ uniform distribution).
I don't know if prob() accept real parameter.

Szunti commented:
You are right Menshin, I just wanted to show you should choose a sympathetic number, it's easier to understand the code that way. And it's really the same.

Szunti commented:
Originally I looked at the code beacuse I thought Aranesp is too deadly.

Incoming5643 commented:
If you'd like a little more work why not deal with this for all prob(rand())'s while you're here?

https://github.com/tgstation/-tg-station/search?utf8=%E2%9C%93&q=%22prob%28rand%22&type=Code

prob(rand(1,100)) = prob(50.5) //but no one's gonna yell at you if you just prob(50) I'm sure
prob(rand(5,50)) = prob(27.5)
prob(rand(50,100)) = prob(75)
prob(rand(FUZZY_CHANCE_LOW,FUZZY_CHANCE_HIGH)) = prob((FUZZY_CHANCE_LOW+FUZZY_CHANCE_HIGH)/2)

Iamgoofball commented:
What is the purpose of doing this? 

I have the randomized prob checks so that I can't just go "okay so 55 of the next 100 cycles is gonna do shit".

With the random checks, it could be 17 out of 100 or 67 out of 100 or 3 out of 100. No set values to get exact numbers with.

AnturK commented:
Goof unless you provide a test case where this helps with byond rng at heavy load thus is literally code obfuscation.

Miauw62 commented:
i'm fairly sure that prob accepts decimals.

Incoming5643 commented:
That's not how the law of large numbers works. Prob(50) could proc for 0 out of a 100 times or 100 out of 100 times, however on average over a number of trials, it will trend towards 50%. 

This is entirely a readability/preference thing, they are effectively the same calculations.

Miauw62 commented:
and what anturk and incoming said, this is just code obfuscation, because it is obvious that the actual chance can be calculated easily, and forcing players to do one-time calculations that never change is bad.

Incoming5643 commented:
To be clear, if you were setting a static result for rand somewhere outside the every tick proc then it WOULD matter, but since its rerolled every time, it just tends to the average

Menshin commented:
>prob(rand(5,50)) = prob(27.5)
prob(rand(50,100)) = prob(75)
prob(rand(FUZZY_CHANCE_LOW,FUZZY_CHANCE_HIGH)) = prob((FUZZY_CHANCE_LOW+FUZZY_CHANCE_HIGH)/2)

This is wrong. How are you establishing this formula ?

> What is the purpose of doing this? 

Nothing really essential, rather a matter of taste:
* showing the 'reasoning' behind the check if(prob(__rand(...)__),
* or, hiding it prob(magic_number)

> I have the randomized prob checks so that I can't just go "okay so 55 of the next 100 cycles is gonna do shit".
With the random checks, it could be 17 out of 100 or 67 out of 100 or 3 out of 100. No set values to get exact numbers with.

Actually, if Byond doesn't implement wacky random functions, Szunti is calculating the total probability of the check succeeding.

Incoming5643 commented:
rand only picks from whole numbers, so rand(1,5) is the same thing as pick(1,2,3,4,5). Over a number of trials the sum of these picks will average towards their mean value, which is just:

(1+2+3+4+5)/5 = 15/5 = 3

Since rand always includes all the numbers between the min and max, the mean is always equivalent the mean of the min and max alone

(1 + 5)/2 = 3

Menshin commented:
>Over a number of trials the sum of these picks will average towards their mean value

We don't need the law of large numbers here.
The universe is small enough to enumerate all its elements and the law of _prob(rand(L...H))_ is exact, determinated by a (rather simple) formula.

Iamgoofball commented:
here'll I'll make a quick test case to figure out who's right here.

Incoming5643 commented:
It really doesn't matter, I'm just saying that unless byond rng is spotty (and people have done large runs that seem to suggest it's solid enough) you can treat rand()'s inside probs as black boxes and replace them with their means. You don't HAVE to, there's no real benefit to it, there's just no real drawback either.

Menshin commented:
Seems like we're agreeing on the whole thing after all...
Either way, it should work just fine and your mean formula may indeed be clearer.

Iirc Byond recently reimplemented the RNG to use Mersenne Twister, so it should be sound enough.

Iamgoofball commented:
http://pastebin.com/CEsepxM6 90 test runs

code used:
http://pastebin.com/yWEJnbeW

proceed with changes to prob(), as it appears it is averaging out to around 50 out of 100 times

Szunti committed:
Replaced prob(rand(A,B)) with prob((A+B)/2).

They are the same probability.
Rounded to integers where the values were literals.

Szunti commented:
So modules/mob/interact.dm has FUZZY_CHANCE_LOW and FUZZY_CHANCE_HIGH always together
in prob(rand(FUZZY_CHANCE_LOW, FUZZY_CHANCE_HIGH)) with one exception, but that is a typo I think. I could remove them and add just one FUZZY_CHANCE constant.

Incoming5643 commented:
I'm gonna guess that since combat() is set to background it's running a lot, so any fuzzy chances there should be called often enough to where you could do that without any outside effect. There ARE fuzzy chances that are only called once, and in those cases you could say they have an effect. You don't really have to go in so deep with this, that system is kind of teeming with these checks.

Incoming5643 commented on code/modules/reagents/Chemistry-Goon-Toxins.dm:
```diff
@@ -218,7 +218,7 @@ datum/reagent/initropidril/on_mob_life(var/mob/living/M as mob)
 	if(prob(33))
 		M.adjustToxLoss(rand(5,25))
-	if(prob(rand(5,10)))
+	if(prob(7))
```
27*

Incoming5643 commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -474,7 +474,7 @@
 							location.add_vomit_floor(src, 1)
 				if("Interstellar Flux")
-					if(prob(rand(50,100))) //50% of the time, it works 100% of the time
+					if(prob(75)))
```
One too many )'s here now

Incoming5643 commented:
Nevermind I was dumb, you're right on that

Szunti commented:
There was a couple of places not shown in github seach. grep is still better.

Szunti commented:
I rewrote the commit to change it everywhere.
